pkg/controller/deployment/recreate.go

```diff
@@ -45,9 +53,16 @@ func (dc *DeploymentController) rolloutRecreate(deployment *extensions.Deploymen
 		return dc.syncRolloutStatus(allRSs, newRS, deployment)
 	}
 
-	// Wait for all old replica set to scale down to zero.
-	if err := dc.waitForInactiveReplicaSets(activeOldRSs); err != nil {
-		return err
+	// Calculate the status of the deployment in order to work with its latest state.
+	// If there are no updatedReplicas running, then we need to wait for all old replicas
+	// to be terminated before proceeding.
+	newStatus := dc.calculateStatus(allRSs, newRS, deployment)
+	// If a secondary worker that watches pod deletions for this Deployment is yet to update
+	// the deployment with an AllPodsTerminatedReason, then we need to stop syncing here. We
+	// will resync once the reason is added in the deployment status.
+	cond := util.GetDeploymentCondition(newStatus, extensions.DeploymentProgressing)
+	if newStatus.UpdatedReplicas == 0 && (cond == nil || cond.Reason != util.AllPodsTerminatedReason) {
```

Comment:
Thinking of using an internal annotation instead of overcomplicating this by using the Progressing condition.